File: legacy/internal/storage/etag.go
package storage

import (
	"crypto/sha256"
	"fmt"
	"sort"
	"strings"
)

// DocumentETag returns the ETag for a document: first 16 hex chars of SHA-256.
func DocumentETag(content []byte) string {
	h := sha256.Sum256(content)
	return fmt.Sprintf("%x", h[:8])
}

// FolderETag returns the ETag for a folder based on its direct children.
// It hashes the sorted name+etag pairs of all children.
func FolderETag(children map[string]string) string {
	names := make([]string, 0, len(children))
	for name := range children {
		names = append(names, name)
	}
	sort.Strings(names)

	var b strings.Builder
	for _, name := range names {
		b.WriteString(name)
		b.WriteString(children[name])
	}

	h := sha256.Sum256([]byte(b.String()))
	return fmt.Sprintf("%x", h[:8])
}

// QuoteETag wraps an ETag value in double quotes for HTTP headers.
func QuoteETag(etag string) string {
	return `"` + etag + `"`
}

// UnquoteETag strips surrounding double quotes from an ETag value.
func UnquoteETag(s string) string {
	if len(s) >= 2 && s[0] == '"' && s[len(s)-1] == '"' {
		return s[1 : len(s)-1]
	}
	return s
}

File: src/Rstash.Model/ETag.cs
using System.Security.Cryptography;
using System.Text;

namespace Rstash.Model;

/// <summary>
/// remoteStorage ETag computation. Ports the Go reference exactly: the ETag is
/// the first 8 bytes of a SHA-256 digest rendered as lowercase hex (16 chars).
/// </summary>
public static class ETag
{
    /// <summary>ETag for a document: first 8 bytes of SHA-256(content), hex.</summary>
    public static string ForDocument(ReadOnlySpan<byte> content)
    {
        Span<byte> hash = stackalloc byte[SHA256.HashSizeInBytes];
        SHA256.HashData(content, hash);
        return Convert.ToHexStringLower(hash[..8]);
    }

    /// <summary>
    /// ETag for a folder, derived from its direct children. Hashes the
    /// ordinal-sorted concatenation of each child's name followed by its ETag,
    /// so the result is stable regardless of enumeration order.
    /// </summary>
    public static string ForFolder(IReadOnlyDictionary<string, string> children)
    {
        var builder = new StringBuilder();
        foreach (var name in children.Keys.OrderBy(static n => n, StringComparer.Ordinal))
        {
            builder.Append(name);
            builder.Append(children[name]);
        }

        Span<byte> hash = stackalloc byte[SHA256.HashSizeInBytes];
        SHA256.HashData(Encoding.UTF8.GetBytes(builder.ToString()), hash);
        return Convert.ToHexStringLower(hash[..8]);
    }

    /// <summary>Wraps an ETag value in double quotes for HTTP headers.</summary>
    public static string Quote(string etag) => $"\"{etag}\"";

    /// <summary>Strips surrounding double quotes from an ETag header value.</summary>
    public static string Unquote(string value) =>
        value.Length >= 2 && value[0] == '"' && value[^1] == '"'
            ? value[1..^1]
            : value;
}
